Sources/HTTP/HTTPConnection.swift

```diff
@@ -0,0 +1,620 @@
+//
+//  HTTPConnection.swift
+//  HTTP
+//
+//  Created by Helge Hess on 22.10.17.
+//  Copyright © 2017 ZeeZide GmbH. All rights reserved.
+//
+
+import Foundation
+import Dispatch
+import CHTTPParser
+
+/// A connection to a single HTTP client.
+///
+/// The connection owns a queue which is used to synchronize all access to the
+/// API. Writers and `finishedProcessing` callbacks alike.
+///
+/// Back-pressure is supported. That is, until the body processor of a handler
+/// marks a chunk as processed, the connection suspends reading more data.
+/// Note that some reads might still be in-flight and will be buffered until
+/// the connection is resumed again.
+///
+/// Since HTTP clients can pipeline calls, a connection can run multiple
+/// handlers at the same time. To maintain proper response ordering,
+/// response-writer objects will be 'corked' until the responses ahead of them
+/// are completely written out.
+///
+internal class HTTPConnection : CustomStringConvertible {
+
+  /// The `HTTPServer` which created the connection. If the connection shuts
+  /// down, it will unregister from that server.
+  private  let server  : HTTPServer
+  
+  /// Serialized access to the state of the connection. Everything needs to
+  /// happen on this queue. It is the `DispatchQueue.main` for a handler.
+  internal let queue   : DispatchQueue
+
+  /// The socket.
+  private  let channel : DispatchIO
+  
+  /// User-level closure responsible for handling any incoming requests. Passed
+  /// in from `HTTPServer`. The connection maintains a strong reference to it,
+  /// until it is gone.
+  private  let requestHandler : HTTPRequestHandler
+  
+  /// The body handler as returned by the `HTTPRequestHandler`. If the handler
+  /// returns `.discard`, this will be nil and all content will be dropped.
+  /// At any time, there can only be one `bodyHandler`.
+  private  var bodyHandler    : HTTPBodyHandler?
+  
+  /// The array of active writers. There can be many due to HTTP pipelining.
+  /// I.e. a new request can arrive while the previous request is still being
+  /// delivered to the client.
+  /// HTTP/1 requires strict ordering of responses. That is, the first response
+  /// must be fully sent before the next can be sent. To support that, inactive
+  /// writers are 'corked' and will be flushed when they are due.
+  private var writers = [ HTTPResponseWriter ]()
+    // FIXME: make this a linked list
+  
+  internal init(fd             : Int32,
```

Comment:
Nit: usually use newline before first arg when splitting args up on multiple lines

Reply:
This actually sounds a little weird, is there a style guide or standard lib example which does it like this?

Reply:
Hm - yeah maybe this is just a thing that we do in my company's internal style guide.  

I actually don't even know, is there an Apple approved style guide?  Couldn't find one with a quick search.

Reply:
You should find out if you propose such changes ;->

Reply:
https://github.com/apple/swift/blob/master/stdlib/public/core/BidirectionalCollection.swift#L166

Reply:
I guess they do this only if the line would have been too long without, in which case they start the arguments on the next line.

I can go check to see if other projects are adhering to some style guide somewhere.  Eventually we will probably run into a bunch of different tiny things like this.

Reply:
This looks really ugly, but fair enough. On the pro side, it uses a proper 2-space indent :-> Maybe we should have an issue to track what style we want to do.

Reply:
👍 .  Will file an issue.

Reply:
Wasn't really arguing with `we do this in my company's internal style guide`.  Only that I'm used to that style guide so I made the mistake of marking this as wrong here.